src/protocols/ec2-query.ts

```diff
@@ -1,21 +1,238 @@
+        shape.member?.locationName ?? shape.member?.queryName ?? "member";
+      const flattened = shape.member?.flattened || shape.flattened;
+      const arrNode = flattened ? node : node?.[memberName];
+      const items = Array.isArray(arrNode)
+        ? arrNode
+        : arrNode != null
+          ? [arrNode]
+          : [];
+      return items.map((it: any) => fromXml(shapes, shape.member!.target, it));
+    }
+
+    case "map": {
+      const entry = shape.flattened ? node : node?.entry;
+      const items = Array.isArray(entry) ? entry : entry != null ? [entry] : [];
+      const out: any = {};
+      for (const it of items) {
+        const k = fromXml(shapes, shape.key!.target, it.key);
+        const v = fromXml(shapes, shape.value!.target, it.value);
+        out[k] = v;
+      }
+      return out;
+    }
+
+    case "timestamp": {
+      if (typeof node === "number") return new Date(node * 1000).toISOString();
+      if (/^\d+$/.test(String(node)))
+        return new Date(Number(node) * 1000).toISOString();
+      const d = new Date(node);
+      return Number.isNaN(+d) ? node : d.toISOString();
+    }
+
+    case "boolean":
+      return String(node) === "true";
+
+    case "integer":
+      return Number(node);
+
+    case "blob":
+      return typeof node === "string" ? node : node?.toString?.("base64");
+
+    default:
+      return node?.toString?.() ?? node;
+  }
+}
+
+function findResponseWrapperName(doc: any): string {
+  const keys = Object.keys(doc);
+  return keys.find((key) => key.endsWith("Response")) || keys[0];
+}
 
 export class Ec2QueryHandler implements ProtocolHandler {
   readonly name = "ec2Query";
-  readonly contentType = "application/x-www-form-urlencoded";
+  readonly contentType = "application/x-www-form-urlencoded; charset=utf-8";
 
   buildRequest(
     input: unknown,
     action: string,
     _metadata: ServiceMetadata,
   ): string {
+    // For now, use fallback synchronous implementation
+    // The enhanced version with metadata will be used when available
     const params = new URLSearchParams();
     params.append("Action", action);
-    params.append("Version", "2016-11-15"); // EC2 API version
+    params.append("Version", "2016-11-15");
 
-    // Flatten the input object into query parameters
-    this.flattenObject(input, "", params);
+    // Try to use enhanced flattening if metadata is available synchronously
+    if (ec2ModelMeta) {
+      const op = ec2ModelMeta.operations[action];
+      if (op) {
+        const enhancedParams: Record<string, string> = {
+          Action: op.name,
+          Version: ec2ModelMeta.version,
+        };
+
+        if (op.input && input) {
+          toParams(ec2ModelMeta.shapes, op.input, input, "", enhancedParams);
+        }
+
+        const usp = new URLSearchParams();
+        for (const [k, v] of Object.entries(enhancedParams)) usp.append(k, v);
+        return usp.toString();
+      }
+    } else {
+      //FIXME: should throw error
+    }
```

Comment:
What's the story here?

Reply:
resolved with change to error handling above